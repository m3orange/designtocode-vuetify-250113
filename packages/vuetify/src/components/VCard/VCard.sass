// Imports
@import './index'

.v-card
  background: $card-background
  color: $card-color
  position: relative

  &::before,
  &::after
    border-radius: inherit
    bottom: 0
    content: ''
    display: block
    left: 0
    pointer-events: none
    position: absolute
    right: 0
    top: 0
    transition-property: box-shadow, opacity
    transition: 0.28s $standard-easing

  &::before
    @include elevation($card-elevation)

  @include border($card-border...)
  @include position($card-positions)
  @include rounded($card-border-radius)

  &--border
    border-width: $card-border-thin-width

  &--disabled
    pointer-events: none
    user-select: none

    > [class*='v-card-']
      opacity: $card-disabled-opacity

  &--hover
    cursor: pointer

    &::before
      opacity: 1
      z-index: -1

    &::after
      z-index: 1
      opacity: 0

      @include elevation($card-hover-elevation)

    &:hover::after
      opacity: 1

    &:hover::before
      opacity: 0

  &--tile
    @include rounded($card-tile-border-radius)

.v-card-title
  align-items: center
  display: flex
  font-size: $card-title-font-size
  font-weight: $card-title-font-weight
  letter-spacing: $card-title-letter-spacing
  text-transform: $card-title-text-transform

  @include card-line-height-densities($card-title-densities)

.v-card-subtitle
  align-items: center
  color: $card-subtitle-color
  display: flex
  font-size: $card-subtitle-font-size
  font-weight: $card-subtitle-font-weight
  letter-spacing: $card-subtitle-letter-spacing
  text-transform: $card-subtitle-text-transform

  @include card-line-height-densities($card-subtitle-densities)

.v-card-text
  color: $card-text-color
  font-size: $card-text-font-size
  font-weight: $card-text-font-weight
  letter-spacing: $card-text-letter-spacing
  padding: $card-text-padding
  text-transform: $card-text-text-transform

  @include card-line-height-densities($card-text-densities)

.v-card-actions
  align-items: center
  display: flex
  flex: 1 1 auto
  min-height: $card-actions-min-height
  padding: $card-actions-padding

.v-card-media
  overflow: hidden

  &:first-child
    .v-card &
      border-top-left-radius: inherit
      border-top-right-radius: inherit

  &:last-child
    .v-card &
      border-bottom-left-radius: inherit
      border-bottom-right-radius: inherit

.v-card-img
  border-radius: inherit
  display: flex
  height: 100%
  left: 0
  overflow: hidden
  position: absolute
  top: 0
  width: 100%
  z-index: -1

.v-card-avatar
  @include ltr()
    margin-right: $card-avatar-margin-start

  @include rtl()
    margin-left: $card-avatar-margin-end

.v-card-header
  flex: 1 1 auto

.v-card-item
  align-items: center
  display: flex
  padding: $card-item-padding
  height: $card-item-height

  .v-card--density-comfortable &
    height: $card-item-comfortable-height

  .v-card--density-compact &
    height: $card-item-compact-height
