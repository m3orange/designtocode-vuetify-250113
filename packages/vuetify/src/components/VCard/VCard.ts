// Styles
import '../../stylus/components/_cards.styl'

// Extensions
import VSheet from '../VSheet'

// Extensions
import VHover from '../VHover'

// Mixins
import Routable from '../../mixins/routable'

// Helpers
import mixins from '../../util/mixins'

// Types
import { VNode } from 'vue'

/* @vue/component */
export default mixins(
  Routable,
  VSheet
).extend({
  name: 'v-card',

  props: {
    elevation: {
      type: [Number, String],
      default: 2
    },
    flat: Boolean,
    hover: [Boolean, String, Number],
    img: String,
    raised: Boolean
  },

  data: () => ({
    isHovered: false
  }),

  computed: {
    classes (): object {
      return {
        'v-card': true,
        'v-card--hover': this.isHoverable,
        ...VSheet.options.computed.classes.call(this)
      }
    },
    computedElevation (): number | string {
      if (this.isHovered) {
        return typeof this.hover === 'boolean' ? 8 : Number(this.hover)
      }
      if (this.flat) return 0
      if (this.raised) return 3

      return (VSheet.options.computed as any).computedElevation.call(this)
    },
    isHoverable (): boolean {
      return this.hover != null && this.hover !== false
    },
    styles (): object {
      const style = {
        ...VSheet.options.computed.styles.call(this)
      }

      if (this.img) {
        style.background = `url("${this.img}") center center / cover no-repeat`
      }

      return style
    }
  },

  render (h): VNode {
    const { tag, data } = this.generateRouteLink(this.classes)

    data.style = this.styles

    const card = h(tag, this.setBackgroundColor(this.color, data), this.$slots.default)

    if (!this.isHoverable) return card

    return h(VHover, {
      props: { value: this.isHovered },
      on: {
        input: (val: boolean) => {
          this.isHovered = val
        }
      }
    }, [card])
  }
})
