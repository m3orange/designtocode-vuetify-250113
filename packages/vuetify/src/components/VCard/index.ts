export { default as VCard } from './VCard'
export { default as VCardActions } from './VCardActions'
export { default as VCardImg } from './VCardImg'
export { default as VCardSubtitle } from './VCardSubtitle'
export { default as VCardText } from './VCardText'
export { default as VCardTitle } from './VCardTitle'
