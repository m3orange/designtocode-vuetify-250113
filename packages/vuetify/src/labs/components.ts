export * from './VConfirmEdit'
export * from './VCalendar'
export * from './VFab'
export * from './VPicker'
export * from './VSparkline'
export * from './VSpeedDial'
export * from './VEmptyState'
