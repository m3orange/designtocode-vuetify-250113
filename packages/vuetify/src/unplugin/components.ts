// Types
import type { ComponentResolver } from 'unplugin-vue-components/types'
import type {
  ComponentName,
  DirectiveName,
  ImportComponents,
  ImportMaps,
  LabComponentName,
} from './types'
import { resolveVuetifyImportMap, resolveVuetifyImportMaps } from './utils'

export interface VuetifyComponentResolverOptions {
  /**
   * Include labs components?.
   *
   * @default false
   */
  labs?: boolean
  /**
   * Components to exclude.
   */
  exclude?: (ComponentName | LabComponentName)[]
  /**
   * Paths to locate Vuetify package.
   *
   * @default [process.cwd()]
   */
  paths?: string[]
}

export interface VuetifyDirectivesResolverOptions {
  /**
   * Prefix Vuetify directives (to allow use other directives with the same name):
   * - when prefix set to `true` will use `Vuetify` => `v-vuetify-<directive>: `v-vuetify-ripple`.
   */
  prefix?: true
  /**
   * Directives to exclude.
   */
  exclude?: DirectiveName[]
  /**
   * Paths to locate Vuetify package.
   *
   * @default [process.cwd()]
   */
  paths?: string[]
}

export interface VuetifyVueResolverOptions extends Omit<VuetifyComponentResolverOptions, 'exclude'> {
  /**
   * Prefix Vuetify directives (to allow use other directives with the same name):
   * - when prefix set to `true` will use `Vuetify` => `v-vuetify-<directive>: `v-vuetify-ripple`.
   */
  prefixDirectives?: true
  /**
   * Directives to exclude.
   */
  excludeDirectives?: DirectiveName[]
  /**
   * Components to exclude.
   */
  excludeComponents?: (ComponentName | LabComponentName)[]
}

export function VuetifyVueResolver (options: VuetifyVueResolverOptions = {}) {
  const {
    paths,
    excludeDirectives,
    labs,
    excludeComponents,
    prefixDirectives,
  } = options

  const [componentsPromise, directivesPromise] = resolveVuetifyImportMaps(paths)

  const directives = createDirectivesResolver(
    componentsPromise,
    { exclude: excludeDirectives, paths, prefix: prefixDirectives }
  )
  const components = createComponentsResolver(
    [componentsPromise, directivesPromise],
    { exclude: excludeComponents, labs, paths }
  )

  return {
    VuetifyDirectiveResolver: directives,
    VuetifyComponentResolver: components,
  }
}

export function VuetifyDirectiveResolver (options: VuetifyDirectivesResolverOptions = {}) {
  return createDirectivesResolver(resolveVuetifyImportMap(options.paths), options)
}

export function VuetifyComponentResolver (options: VuetifyComponentResolverOptions = {}) {
  return createComponentsResolver(resolveVuetifyImportMaps(options.paths), options)
}

function createComponentsResolver (
  promises: ImportMaps,
  options: VuetifyComponentResolverOptions
) {
  const { exclude, labs } = options
  return <ComponentResolver>{
    type: 'component',
    resolve: async name => {
      if (exclude?.some(e => e === name)) return undefined
      const [components, labsComponents] = await Promise.all(promises)
      const component = name in components.components
        ? components.components[name]
        : labs && name in labsComponents
          ? labsComponents[name]
          : undefined

      if (!component) return undefined
      return {
        name,
        from: `vuetify/lib/${component.from}`,
      }
    },
  }
}

function createDirectivesResolver (
  promise: Promise<ImportComponents>,
  options: VuetifyDirectivesResolverOptions
) {
  const { exclude, prefix } = options
  // Vue will transform v-<directive> to _resolveDirective('<directive>')
  // If prefix enabled, Vue will transform v-vuetify-<directive> to _resolveDirective('vuetify-<directive>')
  // unplugin-vue-components will provide the correct import when calling resolve: PascalCase(<directive>)
  // If prefix enabled, unplugin-vue-components will provide PascalCase(vuetify-<directive>)
  return <ComponentResolver>{
    type: 'directive',
    resolve: async resolvedName => {
      let name = resolvedName
      if (prefix) {
        if (!name.startsWith('Vuetify')) {
          return undefined
        }
        name = name.slice('Vuetify'.length)
      }
      if (exclude?.some(e => e === name)) return undefined
      const { directives } = await promise
      const directive = directives.includes(name as DirectiveName)
      if (!directive) return undefined

      return {
        name,
        as: prefix ? resolvedName : undefined,
        from: `vuetify/directives`,
      }
    },
  }
}
