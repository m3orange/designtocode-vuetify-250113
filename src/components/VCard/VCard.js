require('../../stylus/components/_cards.styl')

import Colorable from '../../mixins/colorable'
import Routable from '../../mixins/routable'

export default {
  name: 'v-card',

  mixins: [Colorable, Routable],

  props: {
    flat: Boolean,
    height: {
      type: String,
      default: 'auto'
    },
    hover: Boolean,
    img: String,
    raised: Boolean,
    tag: {
      type: String,
      default: 'div'
    },
    tile: Boolean
  },

  computed: {
    classes () {
      return {
        'card': true,
        'card--flat': this.flat,
        'card--horizontal': this.horizontal,
        'card--hover': this.hover,
        'card--raised': this.raised,
        'card--tile': this.tile,
        'theme--light': this.light,
        'theme--dark': this.dark
      }
    },
    styles () {
      const style = {
        height: isNaN(this.height) ? this.height : `${this.height}px`
      }

      if (this.img) {
        style.background = `url("${this.img}") center center / cover no-repeat`
      }

      return style
    }
  },

  render (h) {
    const { tag, data } = this.generateRouteLink()

    data.class = this._computedClasses
    data.style = this.styles

    return h(tag, data, this.$slots.default)
  }
}
