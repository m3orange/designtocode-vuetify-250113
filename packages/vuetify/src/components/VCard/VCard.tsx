// Styles
import './VCard.sass'

// Components
import { VCardActions, VCardImg, VCardSubtitle, VCardText, VCardTitle } from './'

// Composables
import { makeBorderProps, useBorder } from '@/composables/border'
import { makeDimensionProps, useDimension } from '@/composables/dimensions'
import { makeElevationProps, useElevation } from '@/composables/elevation'
import { makePositionProps, usePosition } from '@/composables/position'
import { makeRoundedProps, useRounded } from '@/composables/rounded'
import { makeTagProps } from '@/composables/tag'
import { useBackgroundColor } from '@/composables/color'
import { useTheme } from '@/composables/theme'

// Utilities
import { defineComponent, toRef } from 'vue'
import { makeProps } from '@/util'

export default defineComponent({
  name: 'VCard',

  props: makeProps({
    color: String,
    flat: Boolean,
    hover: Boolean,
    image: String,
    raised: Boolean,
    subtitle: String,
    text: String,
    title: String,
    ...makeBorderProps(),
    ...makeDimensionProps(),
    ...makeElevationProps(),
    ...makePositionProps(),
    ...makeRoundedProps(),
    ...makeTagProps(),
  }),

  setup (props, { slots }) {
    const { themeClasses } = useTheme()
    const { backgroundColorClasses, backgroundColorStyles } = useBackgroundColor(toRef(props, 'color'))
    const { borderClasses } = useBorder(props, 'v-card')
    const { dimensionStyles } = useDimension(props)
    const { elevationClasses } = useElevation(props)
    const { positionClasses, positionStyles } = usePosition(props, 'v-card')
    const { roundedClasses } = useRounded(props, 'v-card')

    return () => {
      const hasImage = (slots.image || props.image)

      return (
        <props.tag
          class={[
            'v-card',
            {
              'v-card--flat': props.flat,
              'v-card--hover': props.hover,
              // 'v-card--link': props.isClickable,
              // 'v-card--loading': props.loading,
              // 'v-card--disabled': props.disabled,
              'v-card--raised': props.raised,
            },
            themeClasses.value,
            backgroundColorClasses.value,
            borderClasses.value,
            elevationClasses.value,
            positionClasses.value,
            roundedClasses.value,
          ]}
          style={[
            backgroundColorStyles.value,
            dimensionStyles.value,
            positionStyles.value,
          ]}
        >
          { hasImage && (
            <VCardImg>
              { slots.image
                ? slots.image?.({ src: props.image })
                : (<img src={ props.image } />)
              }
            </VCardImg>
          ) }

          { props.title && (
            <VCardTitle>{props.title}</VCardTitle>
          ) }

          { props.subtitle && (
            <VCardSubtitle>{props.subtitle}</VCardSubtitle>
          ) }

          { props.text && (
            <VCardText>{props.text}</VCardText>
          ) }

          { slots.default?.() }

          { slots.actions && (
            <VCardActions>{slots.actions()}</VCardActions>
          ) }
        </props.tag>
      )
    }
  },
})
