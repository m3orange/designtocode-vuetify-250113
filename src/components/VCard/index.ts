import { createSimpleFunctional } from '../../util/helpers'
import VCard from './VCard'
import VCardMedia from './VCardMedia'
import VCardTitle from './VCardTitle'
import Vue from 'vue'

const VCardActions = Vue.extend(createSimpleFunctional('v-card__actions'))
const VCardText = Vue.extend(createSimpleFunctional('v-card__text'))

export { VCard, VCardMedia, VCardTitle, VCardActions, VCardText }

VCard.$_vuetify_subcomponents = {
  VCardMedia,
  VCardTitle,
  VCardActions,
  VCardText
}

export default VCard
