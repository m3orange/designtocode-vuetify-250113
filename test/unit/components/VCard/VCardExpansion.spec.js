import { test } from '@/test'
import { VCardExpansion } from '@/components/VCard'

test('VCardExpansion.js', ({ mount }) => {
  it('should render dropdown class', () => {
    const wrapper = mount(VCardExpansion, {
      slots: {
        default: [{
          render: h => h('div', {
            staticClass: 'fake-text'
          })
        }]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render with cover class', () => {
    const wrapper = mount(VCardExpansion, {
      propsData: {
        cover: true
      },
      slots: {
        default: [{
          render: h => h('div', {
            staticClass: 'fake-text'
          })
        }]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should warn if nothing is provided in slot', () => {
    const wrapper = mount(VCardExpansion)
    const warningTip = '[Vuetify] no text was provided in "v-card-expansion"'

    expect(warningTip).toHaveBeenTipped()
  })
})
