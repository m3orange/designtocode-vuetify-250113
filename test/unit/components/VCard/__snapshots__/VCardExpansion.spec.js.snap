// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VCardExpansion.js should render dropdown class 1`] = `

<div class="v-card__expansion v-card__expansion--dropdown">
  <div class="fake-text">
  </div>
</div>

`;

exports[`VCardExpansion.js should render with cover class 1`] = `

<div class="v-card__expansion v-card__expansion--cover">
  <div class="fake-text">
  </div>
</div>

`;
